lib/presto-client/index.js

```diff
@@ -220,167 +235,212 @@ Client.prototype.statementResource = function(opts) {
+            }
 
-                if (state_callback && (last_state !== response.stats.state || enable_verbose_state_callback)) {
-                    state_callback(null, response.id, response.stats);
-                    last_state = response.stats.state;
+            if (error || (response && response.error)) {
+                clear_timeout();
+                if (first_request) {
+                    var message = "execution error" + (response && response.length > 0 ? ":" + response : "");
+                    if (response && response.error && response.error.message)
+                        message = response.error.message;
+                    error_callback({message, error: (error || response.error), code});
+                } else {
+                    error_callback(error || response.error);
                 }
+                return;
+            }
 
-                if (columns_callback && response.columns && !columns) {
-                    columns = response.columns;
-                    columns_callback(null, columns);
-                }
+            if (first_request && (!response.id || !response.nextUri || !response.infoUri)) {
+                clear_timeout();
+                var error_message = null;
+                if (!response.id)
+                    error_message = "query id missing in response for POST /v1/statement";
+                else if (!response.nextUri)
+                    error_message = "nextUri missing in response for POST /v1/statement";
+                else if (!response.infoUri)
+                    error_message = "infoUri missing in response for POST /v1/statement";
+                error_callback({message: error_message, data: response});
+                return;
+            }
 
-                var fetchNextWithTimeout = function(uri, checkInterval) {
-                    setTimeout(function(){ fetch_next(uri); }, checkInterval);
-                };
-
-                /* presto-main/src/main/java/com/facebook/presto/execution/QueryState.java
-                 * QUEUED, PLANNING, STARTING, RUNNING, FINISHED, CANCELED, FAILED
-                 */
-                if (response.stats.state === 'QUEUED'
-                    || response.stats.state === 'PLANNING'
-                    || response.stats.state === 'STARTING'
-                    || response.stats.state === 'RUNNING' && !response.data) {
-                    fetchNextWithTimeout(response.nextUri, client.checkInterval);
-                    return;
-                }
+            first_request = false;
 
-                if (data_callback && response.data) {
-                    data_callback(null, response.data, response.columns, response.stats);
-                }
+            if (state_callback && (last_state !== response.stats.state || enable_verbose_state_callback)) {
+                state_callback(null, response.id, response.stats);
+                last_state = response.stats.state;
+                query_id = response.id;
```

Comment:
I'm sorry that I missed this at the last review, but should this be out of the `if (state_callback &&...)`?
`query_id` should be set even when `state_callback` is not set.

Reply:
Done in 43a555b26eca3f977f44b1ce9600896bc69c1bd0. I added a `|| query.id` in the off chance that the response does not contain an `id`, we don't lose the previously set one, though the value should be stable over the lifetime of the request.